lib/bundle/installer.rb

```diff
@@ -5,47 +5,121 @@ module Installer
+            end
+      InstallerTask.new cls, arg, verb, entry
+    end
+
+    def report_result(task, result)
+      case result
+      when :success
+        puts Formatter.success(task.header)
+      when :skipped
+        puts "Using #{task.entry.name}"
+      when :failed
+        puts Formatter.error("#{task.header} has failed!")
+      else
+        puts Formatter.error("#{task.header} in unknown state: #{result}")
+      end
+    end
+
+    def report_error(task, error)
+      puts Formatter.error("#{task.header} raised an exception: #{error}")
+    end
+
+    def execute_task(task)
+      begin
+        result = task.cls.install(*task.arg)
+        report_result(task, result)
+        { result => 1 }
+      rescue => error
+        report_error(task, error)
+        { errored: { task.entry.name => error } }
+      end
+    end
+
+    def install_entries(entries)
+      initial_state = {
+        success: 0,
+        skipped: 0,
+        failed: 0,
+        errored: {}
+      }
+
+      entries.reduce(initial_state) do |state, entry|
+        task = task_for_entry(entry)
+        result = execute_task(task)
+        combine_states(state, result)
+      end
+    end
+
+    def combine_states(state, result)
+      state.merge(result) do |_, old, new|
+        case old
+        when Integer # for the counts
+          old + new
+        when Hash # for statuses with info
+          old.merge(new)
         else
-          puts Formatter.error("#{verb} #{entry.name} has failed!")
-          failure += 1
+          raise "Unexpected class in state: #{old.class}"
         end
       end
+    end
 
-      if failure.zero?
-        puts Formatter.success("Homebrew Bundle complete! #{success} Brewfile #{Bundle::Dsl.pluralize_dependency(success)} now installed.")
+    def report_success_and_failure(end_state)
+      failures = end_state[:failed]
+      if failures.zero?
+        succeeded = end_state[:success] || 0
+        skipped = end_state[:skipped] || 0
+        installed = succeeded + skipped
+        puts Formatter.success("Homebrew Bundle complete! #{installed} Brewfile #{Bundle::Dsl.pluralize_dependency(installed)} now installed.")
       else
-        puts Formatter.error("Homebrew Bundle failed! #{failure} Brewfile #{Bundle::Dsl.pluralize_dependency(failure)} failed to install.")
+        puts Formatter.error("Homebrew Bundle failed! #{failures} Brewfile #{Bundle::Dsl.pluralize_dependency(failures)} failed to install.")
+      end
+    end
+
+    def report_errors(errored_entries)
+      error_count = errored_entries.size
+      words = Bundle::Dsl.pluralize_dependency(error_count)
+      error_text = "Homebrew Bundle encountered some errors. #{error_count} Brewfile #{words} failed badly:"
+      puts Formatter.error(error_text)
+      errored_entries.each do |entry, error|
+        puts Formatter.error("\t#{entry}\t => \t#{error}")
       end
+    end
 
-      failure.zero?
+    InstallerTask = Struct.new :cls, :arg, :verb, :entry do
```

Comment:
I don't understand what's going on here.

Reply:
Struct is a convenient way to create a basic class and give it simple functionality. This is equivalent to

```ruby
class InstallerTask
  attr_reader :cls, :arg, :verb, :entry

  def initialize cls, arg, verb, entry
    @cls = cls
    @arg = arg
    @verb = verb
    @entry = entry
  end

  #… other methods inside the block passed to Struct.new

end
```

I chose to go this route because keeping that info in a hash was getting verbose.